File: frontend/app/schemas/pipeline.ts
import { z } from 'zod'

export const cardKindEnum = z.enum(['free', 'contact', 'conversation'])
export type CardKind = z.infer<typeof cardKindEnum>

export const linkedEntityTypeEnum = z.enum(['contact', 'conversation'])
export type LinkedEntityType = z.infer<typeof linkedEntityTypeEnum>

export const terminalKindEnum = z.enum(['won', 'lost', 'resolved'])
export type TerminalKind = z.infer<typeof terminalKindEnum>

const hexColor = z.string().regex(/^#[0-9A-Fa-f]{6}$/, 'Cor hex inválida')

export const stageTemplateSchema = z.object({
  name: z.string().min(1).max(255),
  color: hexColor,
  is_terminal: z.boolean().optional(),
  terminal_kind: terminalKindEnum.optional()
})
export type StageTemplate = z.infer<typeof stageTemplateSchema>

export const pipelineTemplateSchema = z.object({
  key: z.string(),
  name: z.string(),
  description: z.string(),
  icon: z.string(),
  color: z.string(),
  card_kind: cardKindEnum,
  stages: z.array(stageTemplateSchema)
})
export type PipelineTemplate = z.infer<typeof pipelineTemplateSchema>

export const stageSchema = z.object({
  id: z.union([z.string(), z.number()]),
  pipelineId: z.union([z.string(), z.number()]),
  name: z.string(),
  position: z.number(),
  color: z.string(),
  isTerminal: z.boolean(),
  terminalKind: terminalKindEnum.nullish(),
  createdAt: z.string(),
  updatedAt: z.string()
})
export type Stage = z.infer<typeof stageSchema>

export const pipelineSchema = z.object({
  id: z.union([z.string(), z.number()]),
  accountId: z.union([z.string(), z.number()]),
  name: z.string(),
  description: z.string().nullish(),
  templateKey: z.string().nullish(),
  cardKind: cardKindEnum,
  icon: z.string().nullish(),
  color: z.string(),
  archivedAt: z.string().nullish(),
  createdBy: z.union([z.string(), z.number()]).nullish(),
  createdAt: z.string(),
  updatedAt: z.string(),
  stages: z.array(stageSchema).optional()
})
export type Pipeline = z.infer<typeof pipelineSchema>

export const cardSchema = z.object({
  id: z.union([z.string(), z.number()]),
  pipelineId: z.union([z.string(), z.number()]),
  stageId: z.union([z.string(), z.number()]),
  position: z.number(),
  title: z.string(),
  description: z.string().nullish(),
  valueCents: z.number().nullish(),
  valueCurrency: z.string().nullish(),
  dueDate: z.string().nullish(),
  customAttrs: z.record(z.string(), z.unknown()).default({}),
  linkedEntityType: linkedEntityTypeEnum.nullish(),
  linkedEntityId: z.union([z.string(), z.number()]).nullish(),
  assigneeUserIds: z.array(z.union([z.string(), z.number()])).default([]),
  labelIds: z.array(z.union([z.string(), z.number()])).default([]),
  createdBy: z.union([z.string(), z.number()]).nullish(),
  createdAt: z.string(),
  updatedAt: z.string()
})
export type Card = z.infer<typeof cardSchema>

export const createPipelineSchema = z.object({
  name: z.string().min(1, 'Nome obrigatório').max(255),
  description: z.string().max(2000).optional(),
  template_key: z.string().optional(),
  card_kind: cardKindEnum.optional(),
  icon: z.string().optional(),
  color: hexColor.optional(),
  stages: z.array(stageTemplateSchema).optional()
})
export type CreatePipelineForm = z.infer<typeof createPipelineSchema>

export const updatePipelineSchema = z.object({
  name: z.string().min(1).max(255).optional(),
  description: z.string().max(2000).nullable().optional(),
  icon: z.string().optional(),
  color: hexColor.optional()
})
export type UpdatePipelineForm = z.infer<typeof updatePipelineSchema>

export const createStageSchema = z.object({
  name: z.string().min(1, 'Nome obrigatório').max(255),
  color: hexColor.optional(),
  is_terminal: z.boolean().optional(),
  terminal_kind: terminalKindEnum.optional()
})
export type CreateStageForm = z.infer<typeof createStageSchema>

export const updateStageSchema = z.object({
  name: z.string().min(1).max(255).optional(),
  color: hexColor.optional(),
  position: z.number().optional(),
  is_terminal: z.boolean().optional(),
  terminal_kind: terminalKindEnum.optional()
})
export type UpdateStageForm = z.infer<typeof updateStageSchema>

const dateOnly = z.string().regex(/^\d{4}-\d{2}-\d{2}$/, 'Data inválida (YYYY-MM-DD)')

export const createCardSchema = z.object({
  stage_id: z.union([z.string(), z.number()]),
  title: z.string().min(1, 'Título obrigatório').max(500),
  description: z.string().optional(),
  value_cents: z.number().int().optional(),
  value_currency: z.string().length(3).optional(),
  due_date: dateOnly.optional(),
  custom_attrs: z.record(z.string(), z.unknown()).optional(),
  linked_entity_type: linkedEntityTypeEnum.optional(),
  linked_entity_id: z.union([z.string(), z.number()]).optional()
})
export type CreateCardForm = z.infer<typeof createCardSchema>

export const updateCardSchema = z.object({
  title: z.string().min(1).max(500).optional(),
  description: z.string().nullable().optional(),
  value_cents: z.number().int().nullable().optional(),
  value_currency: z.string().length(3).nullable().optional(),
  due_date: dateOnly.nullable().optional(),
  custom_attrs: z.record(z.string(), z.unknown()).optional()
})
export type UpdateCardForm = z.infer<typeof updateCardSchema>

export const moveCardSchema = z.object({
  stage_id: z.union([z.string(), z.number()]),
  position: z.number()
})
export type MoveCardForm = z.infer<typeof moveCardSchema>

File: backend/internal/dto/pipeline.go
package dto

import (
	"time"

	"backend/internal/model"
)

const (
	isoFormat                = "2006-01-02T15:04:05Z07:00"
	linkedTypeContact        = "contact"
	linkedTypeConversation   = "conversation"
	cardKindLabelFree        = "free"
	cardKindLabelContact     = "contact"
	cardKindLabelConv        = "conversation"
	terminalKindLabelWon     = "won"
	terminalKindLabelLost    = "lost"
	terminalKindLabelResolve = "resolved"
)

// ===== Templates =====

type PipelineTemplate struct {
	Key         string                  `json:"key"`
	Name        string                  `json:"name"`
	Description string                  `json:"description"`
	Icon        string                  `json:"icon"`
	Color       string                  `json:"color"`
	CardKind    string                  `json:"card_kind"`
	Stages      []PipelineTemplateStage `json:"stages"`
}

type PipelineTemplateStage struct {
	Name         string  `json:"name"`
	Color        string  `json:"color"`
	IsTerminal   bool    `json:"is_terminal,omitempty"`
	TerminalKind *string `json:"terminal_kind,omitempty"`
}

// ===== Pipelines =====

type CreatePipelineReq struct {
	Name        string                  `json:"name" validate:"required,min=1,max=255"`
	Description *string                 `json:"description,omitempty"`
	TemplateKey *string                 `json:"template_key,omitempty"`
	CardKind    *string                 `json:"card_kind,omitempty" validate:"omitempty,oneof=free contact conversation"`
	Icon        *string                 `json:"icon,omitempty"`
	Color       *string                 `json:"color,omitempty" validate:"omitempty,hexcolor"`
	Stages      []PipelineTemplateStage `json:"stages,omitempty"`
}

type UpdatePipelineReq struct {
	Name        *string `json:"name,omitempty" validate:"omitempty,min=1,max=255"`
	Description *string `json:"description,omitempty"`
	Icon        *string `json:"icon,omitempty"`
	Color       *string `json:"color,omitempty" validate:"omitempty,hexcolor"`
}

type PipelineResp struct {
	ID          int64        `json:"id"`
	AccountID   int64        `json:"account_id"`
	Name        string       `json:"name"`
	Description *string      `json:"description,omitempty"`
	TemplateKey *string      `json:"template_key,omitempty"`
	CardKind    string       `json:"card_kind"`
	Icon        *string      `json:"icon,omitempty"`
	Color       string       `json:"color"`
	ArchivedAt  *string      `json:"archived_at,omitempty"`
	CreatedBy   *int64       `json:"created_by,omitempty"`
	CreatedAt   string       `json:"created_at"`
	UpdatedAt   string       `json:"updated_at"`
	Stages      []StageResp  `json:"stages,omitempty"`
}

func PipelineToResp(p *model.Pipeline, stages []model.PipelineStage) PipelineResp {
	resp := PipelineResp{
		ID:          p.ID,
		AccountID:   p.AccountID,
		Name:        p.Name,
		Description: p.Description,
		TemplateKey: p.TemplateKey,
		CardKind:    CardKindToString(p.CardKind),
		Icon:        p.Icon,
		Color:       p.Color,
		CreatedBy:   p.CreatedBy,
		CreatedAt:   p.CreatedAt.Format(isoFormat),
		UpdatedAt:   p.UpdatedAt.Format(isoFormat),
	}
	if p.ArchivedAt != nil {
		s := p.ArchivedAt.Format(isoFormat)
		resp.ArchivedAt = &s
	}
	if stages != nil {
		resp.Stages = StagesToResp(stages)
	}
	return resp
}

func PipelinesToResp(items []model.Pipeline) []PipelineResp {
	out := make([]PipelineResp, len(items))
	for i := range items {
		out[i] = PipelineToResp(&items[i], nil)
	}
	return out
}

// ===== Stages =====

type CreateStageReq struct {
	Name         string  `json:"name" validate:"required,min=1,max=255"`
	Color        *string `json:"color,omitempty" validate:"omitempty,hexcolor"`
	IsTerminal   *bool   `json:"is_terminal,omitempty"`
	TerminalKind *string `json:"terminal_kind,omitempty" validate:"omitempty,oneof=won lost resolved"`
}

type UpdateStageReq struct {
	Name         *string  `json:"name,omitempty" validate:"omitempty,min=1,max=255"`
	Color        *string  `json:"color,omitempty" validate:"omitempty,hexcolor"`
	Position     *float64 `json:"position,omitempty"`
	IsTerminal   *bool    `json:"is_terminal,omitempty"`
	TerminalKind *string  `json:"terminal_kind,omitempty" validate:"omitempty,oneof=won lost resolved"`
}

type StageResp struct {
	ID           int64   `json:"id"`
	PipelineID   int64   `json:"pipeline_id"`
	Name         string  `json:"name"`
	Position     float64 `json:"position"`
	Color        string  `json:"color"`
	IsTerminal   bool    `json:"is_terminal"`
	TerminalKind *string `json:"terminal_kind,omitempty"`
	CreatedAt    string  `json:"created_at"`
	UpdatedAt    string  `json:"updated_at"`
}

func StageToResp(s *model.PipelineStage) StageResp {
	resp := StageResp{
		ID:         s.ID,
		PipelineID: s.PipelineID,
		Name:       s.Name,
		Position:   s.Position,
		Color:      s.Color,
		IsTerminal: s.IsTerminal,
		CreatedAt:  s.CreatedAt.Format(isoFormat),
		UpdatedAt:  s.UpdatedAt.Format(isoFormat),
	}
	if s.TerminalKind != nil {
		k := TerminalKindToString(*s.TerminalKind)
		resp.TerminalKind = &k
	}
	return resp
}

func StagesToResp(items []model.PipelineStage) []StageResp {
	out := make([]StageResp, len(items))
	for i := range items {
		out[i] = StageToResp(&items[i])
	}
	return out
}

// ===== Cards =====

type CreateCardReq struct {
	StageID          int64           `json:"stage_id" validate:"required"`
	Title            string          `json:"title" validate:"required,min=1,max=500"`
	Description      *string         `json:"description,omitempty"`
	ValueCents       *int64          `json:"value_cents,omitempty"`
	ValueCurrency    *string         `json:"value_currency,omitempty" validate:"omitempty,len=3"`
	DueDate          *string         `json:"due_date,omitempty" validate:"omitempty,datetime=2006-01-02"`
	CustomAttrs      map[string]any  `json:"custom_attrs,omitempty"`
	LinkedEntityType *string         `json:"linked_entity_type,omitempty" validate:"omitempty,oneof=contact conversation"`
	LinkedEntityID   *int64          `json:"linked_entity_id,omitempty"`
}

type UpdateCardReq struct {
	Title         *string        `json:"title,omitempty" validate:"omitempty,min=1,max=500"`
	Description   *string        `json:"description,omitempty"`
	ValueCents    *int64         `json:"value_cents,omitempty"`
	ValueCurrency *string        `json:"value_currency,omitempty" validate:"omitempty,len=3"`
	DueDate       *string        `json:"due_date,omitempty" validate:"omitempty,datetime=2006-01-02"`
	CustomAttrs   map[string]any `json:"custom_attrs,omitempty"`
}

type MoveCardReq struct {
	StageID  int64   `json:"stage_id" validate:"required"`
	Position float64 `json:"position"`
}

type AssignAgentReq struct {
	UserID int64 `json:"user_id" validate:"required"`
}

type ApplyCardLabelReq struct {
	LabelID int64 `json:"label_id" validate:"required"`
}

type CardResp struct {
	ID               int64          `json:"id"`
	PipelineID       int64          `json:"pipeline_id"`
	StageID          int64          `json:"stage_id"`
	Position         float64        `json:"position"`
	Title            string         `json:"title"`
	Description      *string        `json:"description,omitempty"`
	ValueCents       *int64         `json:"value_cents,omitempty"`
	ValueCurrency    *string        `json:"value_currency,omitempty"`
	DueDate          *string        `json:"due_date,omitempty"`
	CustomAttrs      map[string]any `json:"custom_attrs"`
	LinkedEntityType *string        `json:"linked_entity_type,omitempty"`
	LinkedEntityID   *int64         `json:"linked_entity_id,omitempty"`
	AssigneeUserIDs  []int64        `json:"assignee_user_ids"`
	LabelIDs         []int64        `json:"label_ids"`
	CreatedBy        *int64         `json:"created_by,omitempty"`
	CreatedAt        string         `json:"created_at"`
	UpdatedAt        string         `json:"updated_at"`
}

func CardToResp(c *model.PipelineCard, assignees []int64, labels []int64) CardResp {
	resp := CardResp{
		ID:              c.ID,
		PipelineID:      c.PipelineID,
		StageID:         c.StageID,
		Position:        c.Position,
		Title:           c.Title,
		Description:     c.Description,
		ValueCents:      c.ValueCents,
		ValueCurrency:   c.ValueCurrency,
		AssigneeUserIDs: assignees,
		LabelIDs:        labels,
		CreatedBy:       c.CreatedBy,
		CreatedAt:       c.CreatedAt.Format(isoFormat),
		UpdatedAt:       c.UpdatedAt.Format(isoFormat),
	}
	if resp.AssigneeUserIDs == nil {
		resp.AssigneeUserIDs = []int64{}
	}
	if resp.LabelIDs == nil {
		resp.LabelIDs = []int64{}
	}
	if c.DueDate != nil {
		s := c.DueDate.Format("2006-01-02")
		resp.DueDate = &s
	}
	if c.LinkedEntityType != nil {
		t := LinkedEntityTypeToString(*c.LinkedEntityType)
		resp.LinkedEntityType = &t
	}
	if c.LinkedEntityID != nil {
		resp.LinkedEntityID = c.LinkedEntityID
	}
	if c.CustomAttrs != "" {
		resp.CustomAttrs = parseCustomAttrs(c.CustomAttrs)
	}
	if resp.CustomAttrs == nil {
		resp.CustomAttrs = map[string]any{}
	}
	return resp
}

// ===== Helpers =====

func CardKindToString(k model.CardKind) string {
	switch k {
	case model.CardKindContact:
		return cardKindLabelContact
	case model.CardKindConversation:
		return cardKindLabelConv
	default:
		return cardKindLabelFree
	}
}

func CardKindFromString(s string) (model.CardKind, bool) {
	switch s {
	case cardKindLabelFree:
		return model.CardKindFree, true
	case cardKindLabelContact:
		return model.CardKindContact, true
	case cardKindLabelConv:
		return model.CardKindConversation, true
	}
	return 0, false
}

func LinkedEntityTypeToString(t model.LinkedEntityType) string {
	if t == model.LinkedEntityConversation {
		return linkedTypeConversation
	}
	return linkedTypeContact
}

func LinkedEntityTypeFromString(s string) (model.LinkedEntityType, bool) {
	switch s {
	case linkedTypeContact:
		return model.LinkedEntityContact, true
	case linkedTypeConversation:
		return model.LinkedEntityConversation, true
	}
	return 0, false
}

func TerminalKindToString(k model.TerminalKind) string {
	switch k {
	case model.TerminalKindLost:
		return terminalKindLabelLost
	case model.TerminalKindResolved:
		return terminalKindLabelResolve
	default:
		return terminalKindLabelWon
	}
}

func TerminalKindFromString(s string) (model.TerminalKind, bool) {
	switch s {
	case terminalKindLabelWon:
		return model.TerminalKindWon, true
	case terminalKindLabelLost:
		return model.TerminalKindLost, true
	case terminalKindLabelResolve:
		return model.TerminalKindResolved, true
	}
	return 0, false
}

func ParseDueDate(s string) (*time.Time, error) {
	if s == "" {
		return nil, nil
	}
	t, err := time.Parse("2006-01-02", s)
	if err != nil {
		return nil, err
	}
	return &t, nil
}
